api/applicationserver_pubsub.proto

```diff
@@ -0,0 +1,138 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+import "github.com/envoyproxy/protoc-gen-validate/validate/validate.proto";
+import "github.com/gogo/protobuf/gogoproto/gogo.proto";
+import "google/api/annotations.proto";
+import "google/protobuf/empty.proto";
+import "google/protobuf/field_mask.proto";
+import "google/protobuf/timestamp.proto";
+import "lorawan-stack/api/identifiers.proto";
+
+package ttn.lorawan.v3;
+
+option go_package = "go.thethings.network/lorawan-stack/pkg/ttnpb";
+
+message ApplicationPubSubIdentifiers {
+  ApplicationIdentifiers application_ids = 1 [(gogoproto.embed) = true, (gogoproto.nullable) = false, (validate.rules).message.required = true];
+  string pubsub_id = 2 [(gogoproto.customname) = "PubSubID", (validate.rules).string = {pattern: "^[a-z0-9](?:[-]?[a-z0-9]){2,}$" , max_len: 36}];
+}
+
+message ApplicationPubSub {
+  ApplicationPubSubIdentifiers ids = 1 [(gogoproto.embed) = true, (gogoproto.nullable) = false, (validate.rules).message.required = true];
+  google.protobuf.Timestamp created_at = 2 [(gogoproto.nullable) = false, (gogoproto.stdtime) = true];
+  google.protobuf.Timestamp updated_at = 3 [(gogoproto.nullable) = false, (gogoproto.stdtime) = true];
+
+  // Provider specific attributes used for authentication and settings.
+  map<string,string> attributes = 4;
+  // The format to use for the body.
+  // Supported values depend on the Application Server configuration.
+  string format = 5;
+
+  // The provider for the PubSub.
+  enum Provider {
+      AWSSNSSQS = 0; // AWS SQS / SNS
+      AZURESB = 1;   // Azure Service Bus
+      GCPPUBSUB = 2; // GCP PubSub
+      KAFKA = 3;     // Apache Kafka
+      NATS = 4;      // NATS
+      RABBIT = 5;    // RabbitMQ
+  }
```

Comment:
I followed https://github.com/TheThingsNetwork/lorawan-stack/issues/830#issuecomment-503461660 since it allows more flexibility regarding platform specific configuration. It is also the case that since we are missing the other implementations we might need new fields / remove fields later, which would violate API stability.

Reply:
Maybe I don't understand, but [that comment](https://github.com/TheThingsNetwork/lorawan-stack/issues/830#issuecomment-503461660) supports my argument instead of the opposite, right?

I mean that we have (sub) messages per platform, and a `oneof` which not only indicates which platform to use, but also per-platform specific fields.

Reply:
I'll convert it to a `oneof` then. We'll gain some type-safety in the process.